packages/sdk-rtl/src/paginator.ts

```diff
@@ -0,0 +1,325 @@
+   */
+  prevPage(): Promise<SDKResponse<TSuccess, TError>>
+}
+
+/** Pagination function call */
+export type PaginateFunc<TSuccess, TError> = () => Promise<
+  SDKResponse<TSuccess, TError>
+>
+
+/**
+ * Parse a link header to extract rels
+ * @param linkHeader to parse
+ *
+ * Several different approaches are discussed at https://stackoverflow.com/questions/8735792/how-to-parse-link-header-from-github-api
+ *
+ * Unfortunately, none of them are a good implementation
+ *
+ */
+export const linkHeaderParser = (linkHeader: string): PageLinks => {
+  const re = /<\s*(.*)\s*>;\s*rel="\s*(.*)\s*"\s*/gm
+  const links = linkHeader.split(',')
+  const obj: PageLinks = {}
+  let arrRes
+
+  links.forEach((link) => {
+    link = link.trim()
+    while ((arrRes = re.exec(link))) {
+      const key = arrRes[2].split(' ')[0].trim().toLocaleLowerCase()
+      obj[key] = {
+        url: arrRes[1].trim(),
+        rel: key as PageLinkRel,
+        name: arrRes[2].trim(),
+      }
+    }
+  })
+  return obj
+}
+
+/**
+ * Create an API paginator for an endpoint that returns a Link header
+ * @param sdk implementation of IAPIMethods. Can be full SDK or functional auth session
+ * @param func sdk call that includes a pagination header
+ * @param options transport options override to capture and use in paging requests
+ *
+ * @example
+ * ```ts
+ * const paged = new paginate(sdk, (sdk) => search_dashboards(sdk, { limit: 10 })
+ * const dashboards = paged.items
+ * while (paged.has('next')) {
+ *   await paged.next()
+ *   dashboards.push(paged.items)
+ * }
+ * console.log(`${paged.total} dashboards retrieved`)
+ * ... (some code to list all dashboards )
+ *
+ * ```
+ */
+export async function paginate<TSuccess, TError>(
+  sdk: IAPIMethods,
+  func: PaginateFunc<TSuccess, TError>,
+  options?: Partial<ITransportSettings>
+): Promise<IPaginate<TSuccess, TError>> {
+  return await new Paginator<TSuccess, TError>(sdk, func, options).init()
+}
+
+/**
+ * Pagination support class
+ */
+export class Paginator<TSuccess, TError>
+  implements IPaginate<TSuccess, TError>
+{
+  items: TSuccess = [] as unknown as TSuccess
+  links: PageLinks = {}
+  total = -1
+  offset = -1
+  limit = -1
+
+  private transport: BaseTransport
+
+  /**
+   * Create an API paginator
+   * @param sdk functional AuthSession or full SDK implementation
+   * @param func sdk function to call
+   * @param options transport overrides to use for subsequent requests
+   */
+  constructor(
+    public sdk: IAPIMethods,
+    public func: PaginateFunc<TSuccess, TError>,
+    public options?: Partial<ITransportSettings>
+  ) {
+    this.transport = sdk.authSession.transport as BaseTransport
+  }
+
+  private async rawCatch(func: () => any) {
+    let raw: IRawResponse = {} as IRawResponse
+    const saved = this.transport.observer
+    try {
+      this.transport.observer = (response: IRawResponse) => {
+        raw = response
+        return response
+      }
```

Comment:
this is the cleanest way I could come up with capturing the raw request values for processing the `X-Total-Count` and `Link` headers while still allowing the consumer to just make a call to an SDK function.